core/src/main/java/io/questdb/std/DirectLongHashSet.java

```diff
@@ -0,0 +1,286 @@
+
+/**
+ * Direct (off-heap) long hash set without sentinel value conflicts.
+ */
+public class DirectLongHashSet implements Closeable, Mutable, Sinkable {
+    public static final double DEFAULT_LOAD_FACTOR = 0.4;
+    private static final int MIN_CAPACITY = 16;
+    private final double loadFactor;
+    private final int memoryTag;
+    private int capacity;
+    private int free;
+    private boolean hasZero;
+    private int mask;
+    private long memLimit;
+    private long memStart;
+    private int size;
+
+    public DirectLongHashSet(int capacity) {
+        this(capacity, DEFAULT_LOAD_FACTOR, MemoryTag.NATIVE_DEFAULT);
+    }
+
+    public DirectLongHashSet(int capacity, double loadFactor, int memoryTag) {
+        if (loadFactor <= 0d || loadFactor >= 1d) {
+            throw new IllegalArgumentException("0 < loadFactor < 1");
+        }
+
+        this.loadFactor = loadFactor;
+        this.memoryTag = memoryTag;
+        this.capacity = Math.max(Numbers.ceilPow2((int) (capacity / loadFactor)), MIN_CAPACITY);
+        this.mask = this.capacity - 1;
+        this.free = (int) (this.capacity * loadFactor);
+        this.size = 0;
+        this.hasZero = false;
+
+        long sizeBytes = (long) this.capacity * Long.BYTES;
+        this.memStart = Unsafe.malloc(sizeBytes, memoryTag);
+        this.memLimit = memStart + sizeBytes;
+        Vect.memset(memStart, sizeBytes, 0);
+    }
+
+    public boolean add(long key) {
+        if (key == 0) {
+            return addZero();
+        }
+        return addNonZero(key);
+    }
+
+    public int capacity() {
+        return capacity;
+    }
+
+    @Override
+    public void clear() {
+        if (memStart != 0) {
+            Vect.memset(memStart, memLimit - memStart, 0);
+            size = 0;
+            free = (int) (capacity * loadFactor);
+            hasZero = false;
+        }
+    }
+
+    @Override
+    public void close() {
+        if (memStart != 0) {
+            Unsafe.free(memStart, memLimit - memStart, memoryTag);
+            memStart = memLimit = 0;
+            size = 0;
+            free = 0;
+            hasZero = false;
+        }
+    }
+
+    public boolean contains(long key) {
+        if (key == 0) {
+            return hasZero;
+        }
+        return keyIndex(key) < 0;
+    }
+
+    public boolean excludes(long key) {
+        return !contains(key);
+    }
+
+    public int keyIndex(long key) {
+        if (key == 0) {
+            return hasZero ? -1 : 0;
+        }
+
+        long index = Hash.hashLong64(key) & mask;
+        long addr = memStart + (index * Long.BYTES);
+
+        for (; ; ) {
+            long k = Unsafe.getUnsafe().getLong(addr);
+            if (k == 0) {
+                return (int) index;
+            } else if (k == key) {
+                return (int) -(index + 1);
+            }
+            index = (index + 1) & mask;
+            addr = memStart + (index * Long.BYTES);
+        }
+    }
+
+    public int remove(long key) {
```

Comment:
Why do we need the remove method?

Reply:
It’s currently unused, but since it’s part of the std package, it should provide an interface similar to `LongHashSet`(IMHO).

Reply:
It's not a must. We have some collections with similar names that have different APIs. It's a matter of describing the use case in the class javadoc.